dynamiq/nodes/agents/orchestrators/graph.py

```diff
@@ -0,0 +1,319 @@
+from dynamiq.utils.logger import logger
+
+
+class StateNotFoundError(OrchestratorError):
+    """Raised when next state was not found."""
+    pass
+
+
+START = "START"
+END = "END"
+
+
+class GraphOrchestrator(Orchestrator):
+    """
+    Orchestrates the execution of complex tasks using multiple specialized agents.
+
+    This class manages the breakdown of a main objective into subtasks,
+    delegates these subtasks to appropriate agents, and synthesizes the results
+    into a final answer.
+
+    Attributes:
+        manager (ManagerAgent): The managing agent responsible for overseeing the orchestration process.
+        context (Dict[str, Any]): Context of the orchestrator.
+        states (List[State]): List of states within graph orchestrator.
+        initial_state (str): State to start from.
+        objective (Optional[str]): The main objective of the orchestration.
+        max_loops (Optional[int]): Maximum number of transition between states.
+    """
+
+    name: str | None = "GraphOrchestrator"
+    manager: GraphAgentManager
+    initial_state: str = START
+    context: dict[str, Any] = {}
+    states: list[State] = []
+    max_loops: int = 15
+
+    def init_components(self, connection_manager: ConnectionManager | None = None) -> None:
+        """
+        Initialize components of the orchestrator.
+
+        Args:
+            connection_manager (ConnectionManager, optional): The connection manager. Defaults to ConnectionManager.
+        """
+        super().init_components(connection_manager)
+
+        if self.manager.is_postponed_component_init:
+            self.manager.init_components(connection_manager)
+
+        for state in self.states:
+            if state.is_postponed_component_init:
+                state.init_components(connection_manager)
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self._state_by_id = {state.id: state for state in self.states}
+
+        if START not in self._state_by_id:
+            start_state = State(id=START, description="Initial state")
+            self._state_by_id[START] = start_state
+            self.states.append(start_state)
+
+        if END not in self._state_by_id:
+            end_state = State(id=END, description="Final state")
+            self._state_by_id[END] = end_state
+            self.states.append(end_state)
+
+    @property
+    def to_dict_exclude_params(self):
+        return super().to_dict_exclude_params | {"manager": True, "states": True}
+
+    def to_dict(self, **kwargs) -> dict:
+        """Converts the instance to a dictionary.
+
+        Returns:
+            dict: A dictionary representation of the instance.
+        """
+        data = super().to_dict(**kwargs)
+        data["manager"] = self.manager.to_dict(**kwargs)
+        data["states"] = [state.to_dict(**kwargs) for state in self.states]
+        return data
+
+    def add_state(self, state_id: str, tasks: list[Node | Callable]) -> None:
```

Comment:
Let's rename it little bit to smth like `add_state_by_tasks`.
`add_state` is a general and for me, it should looks like `def add_state(self, state: State)`. In such case I could define own name, description and so on.

Reply:
Agree, but it seems that it gets a bit wordy and less user-friendly. I do not see any need to add separate function which accepts State objects. Also it is implemented in the same way in LangGraph.

Reply:
It's ok if we add a little bit more comparing to langraph. How I could add mine `State` to graph as I'm able to init State by my self? Please add this functionality